askama_derive/src/input.rs

```diff
@@ -88,61 +69,76 @@ impl<'a> TemplateInput<'a> {
         match (&source, ext.is_some()) {
             (&Source::Path(_), true) => {
                 panic!("'ext' attribute cannot be used with 'path' attribute")
-            }
+            },
             (&Source::Source(_), false) => {
                 panic!("must include 'ext' attribute when using 'source' attribute")
-            }
-            _ => {}
+            },
+            _ => {},
         }
 
-        let escaping = match escaping {
-            Some(m) => m,
-            None => {
-                let ext = match source {
-                    Source::Path(ref p) => Path::new(p)
-                        .extension()
-                        .map(|s| s.to_str().unwrap())
-                        .unwrap_or(""),
-                    Source::Source(_) => ext.as_ref().unwrap(), // Already panicked if None
-                };
-                if HTML_EXTENSIONS.contains(&ext) {
-                    EscapeMode::Html
-                } else {
-                    EscapeMode::None
-                }
-            }
-        };
-
-        let parent = match ast.data {
-            syn::Data::Struct(syn::DataStruct {
-                fields: syn::Fields::Named(ref fields),
-                ..
-            }) => fields.named.iter().filter_map(|f| {
-                f.ident
-                    .as_ref()
-                    .and_then(|name| if name == "_parent" { Some(&f.ty) } else { None })
-            }),
-            _ => panic!("derive(Template) only works for struct items"),
-        }.next();
-
         let path = match source {
-            Source::Source(_) => match ext {
-                Some(ref v) => PathBuf::from(format!("{}.{}", ast.ident, v)),
-                None => PathBuf::new(),
-            },
-            Source::Path(ref s) => path::find_template_from_path(s, None),
+            Source::Source(_) => PathBuf::from(format!("{}.{}", ast.ident, ext.unwrap())),
+            Source::Path(ref path) => path::find_template_from_path(path, None),
         };
 
         TemplateInput {
             ast,
             source,
             print,
-            escaping,
-            ext,
-            parent,
+            escaping: escaping.unwrap_or(EscapeMode::from(&path)),
+            parent: field_type(ast, "_parent"),
```

Comment:
This was moved to a function called `field_type` because it took me quite some time to figure out that this gets the type of the `_parent` field at all.
This is another example of single-use you mentioned. It's not likely and "efficient" to reuse the `field_type` function but now the intend is clear and it's better than adding a comment or renaming `parent` to `parent_type` IMHO.

Reply:
Yeah, I don't think these changes will make the code easier to read, generally. It would be good to add some documentation to the `TemplateInput` code to clarify the flow, though!